### PR TITLE
Add `nunit` NUnit make target to Makefile.

### LINKED ISSUE #8849: Completely replace OpenRA.Test with OpenRA.Utility.exe or Support It
It seems that `OpenRA.Test` does not provide much functionality and I have issues with running it on my system:
`dmcs -t:library -out:OpenRA.Test.dll OpenRA.Test/OpenRA.Game/*.cs -r:mods/common/OpenRA.Mods.Common.dll -r:OpenRA.Game.exe -r:nunit.framework.dll -r:System.dll -r:System.Core.dll -r:System.Data.dll -r:System.Data.DataSetExtensions.dll -r:System.Drawing.dll -r:System.Xml.dll -r:thirdparty/download/ICSharpCode.SharpZipLib.dll -r:thirdparty/download/FuzzyLogicLibrary.dll -r:thirdparty/download/Mono.Nat.dll -r:thirdparty/download/MaxMind.Db.dll -r:thirdparty/download/MaxMind.GeoIP2.dll -r:thirdparty/download/Eluant.dll && nunit-console OpenRA.Test.dll`
results in:
```
OpenRA.Test/OpenRA.Game/ActorInfoTest.cs(67,9): warning CS0219: The variable `i' is assigned but its value is never used
OpenRA.Test/OpenRA.Game/ActorInfoTest.cs(88,9): warning CS0219: The variable `i' is assigned but its value is never used
Compilation succeeded - 2 warning(s)
NUnit version 2.4.8
Copyright (C) 2002-2007 Charlie Poole.
Copyright (C) 2002-2004 James W. Newkirk, Michael C. Two, Alexei A. Vorontsov.
Copyright (C) 2000-2002 Philip Craig.
All Rights Reserved.

Runtime Environment - 
   OS Version: Unix 4.1.3.2
  CLR Version: 4.0.30319.17020 ( 4.0.2 (Stable 4.0.2.5/c99aa0c Mon Jul 13 18:57:35 UTC 2015) )

Symbol file /tmp/nunit20/ShadowCopyCache/8873_635738788612624820/Tests/assembly/shadow/0f6609dd/157c8393_1a1a8a4e_00000001/OpenRA.Test.dll.mdb doesn't match image (null)

Tests run: 0, Failures: 0, Not run: 0, Time: 0.014 seconds


```
That probably is something with my system, but it is another possible dependency / point of failure.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25195330-completely-replace-openra-test-with-openra-utility-exe-or-support-it?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by atlimit8**
This allows for overrides from the `make nunit ...` command line:
- Use NUNIT_CONSOLE if nunit[2]-console is not in bin search paths
- Use NUNIT_LIBS_PATH if NUnit libs are not in search paths. Include trailing /
- Use NUNIT_LIBS if NUnit libs have different names (such as including a prefix or suffix)

Closes https://github.com/OpenRA/OpenRA/issues/8849.

### CODE REVIEW

**Comment by pchote** [comment]
Please add a check and error message so that older versions of nunit (i.e. the one that ships with the default mono install) fail with a sensible error instead of running no tests and passing.

**Comment by atlimit8** [comment]
@pchote: Added version checking and error reporting.

**Comment by atlimit8** [comment]
Should this be marked as closing #8849?

**Comment by pchote** [comment]
Something still isn't right here:
```
paul@Mjolnir:OpenRA+pr/8861$ make dependencies
[...snip...]
Fetching NUnit from NuGet
Installing 'NUnit 2.6.4'.
Successfully installed 'NUnit 2.6.4'.
[...snip...]
paul@Mjolnir:OpenRA+pr/8861$ make nunit
CSC OpenRA.Test.dll

Checking unit tests...
NUnit version 2.4.8
NUnit version >= 2.6 required
make: *** [nunit] Error 1
```

**Comment by Mailaender** [comment]
> This should be using our local nunit version.

as a backfall for Operating Systems without proper package management.

**Comment by pchote** [comment]
> as a backfall for Operating Systems without proper package management.

Including linux, right? :wink:
We're already downloading parts of NUnit from nuget on all platforms.

**Comment by atlimit8** [comment]
Low priority right now.

**Comment by atlimit8** [comment]
I'm not sure what to do about the `TestResult.xml` file generated.
Should it go somewhere?
Should it just be deleted?

**Comment by atlimit8** [comment]
The `lib` directory is required for `nunit-console.exe` to work.

File: Makefile [hunk 1]
```diff
@@ -211,6 +223,14 @@ check: utility mods
 	@echo "Checking for code style violations in OpenRA.Test..."
 	@mono --debug OpenRA.Utility.exe ra --check-code-style OpenRA.Test
 
+NUNIT_CONSOLE := $(shell test -f nunit-console.exe && echo mono nunit-console.exe || which nunit2-console 2>/dev/null || which nunit3-console 2>/dev/null || which nunit-console 2>/dev/null)
+nunit: test_dll
+	@echo
+	@echo "Checking unit tests..."
+	@if [[ "$(NUNIT_CONSOLE)" = "" ]] ;then echo 'nunit[2|3]-console not found!'; echo 'Was "make dependencies" called or is NUnit installed?'>&2; exit 1; fi
```
Comment (obrakmann) [inline Makefile]: `[[` is a shell-builtin and doesn't work here. Use a single `[` instead.

Comment (obrakmann) [inline Makefile]: The error message should also point the user to `make help`. It wasn't obvious to me why `make nunit` still failed even though I had run `make dependencies`.

File: thirdparty/fetch-thirdparty-deps.sh [hunk 2]
```diff
@@ -78,6 +78,15 @@ if [ ! -f nunit.framework.dll ]; then
 	rm -rf NUnit
 fi
 
+if [ ! -f nunit-console.exe ]; then
+	echo "Fetching NUnit.Runners from NuGet"
+	get NUnit.Runners 2.6.4
+	cp ./NUnit.Runners/tools/nunit-console.exe .
```
Comment (obrakmann) [inline thirdparty/fetch-thirdparty-deps.sh]: You need to `chmod +x` it, too.
```
$ make nunit NUNIT_CONSOLE=thirdparty/download/nunit-console.exe NUNIT_LIBS_PATH=thirdparty/download/

Checking unit tests...
/bin/sh: 1: thirdparty/download/nunit-console.exe: Permission denied
make: execvp: thirdparty/download/nunit-console.exe: Permission denied
make: *** [nunit] Error 127
```

**Comment by atlimit8** [comment]
Changed.
I should have moved my system's `nunit[2]-console` while before. Hopefully, this is everything.

**Comment by obrakmann** [comment]
yup, works now. :+1:

**Comment by Mailaender** [comment]
Really? Both `make nunit NUNIT_CONSOLE=/usr/lib/nunit/nunit-console.exe NUNIT_LIBS_PATH=/usr/lib/nunit/` and `make nunit NUNIT_CONSOLE=thirdparty/download/nunit-console.exe NUNIT_LIBS_PATH=thirdparty/download` give me "can't execute file" errors for no apparent reason.

**Comment by pchote** [comment]
@Mailaender: I found that I needed to re-`make dependencies` for this to work:
```
paul@Mjolnir:OpenRA+pr/8861$ make nunit
CSC OpenRA.Test.dll

Checking unit tests...
NUnit version 2.4.8
NUnit version >= 2.6 required
make: *** [nunit] Error 1
paul@Mjolnir:OpenRA+pr/8861$ make dependencies
Fetching NUnit.Runners from NuGet
Installing 'NUnit.Runners 2.6.4'.
Successfully installed 'NUnit.Runners 2.6.4'.
paul@Mjolnir:OpenRA+pr/8861$ make nunit
[snip]
Tests run: 11, Errors: 0, Failures: 0, Inconclusive: 0, Time: 0.1386344 seconds
  Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
```

**Comment by pchote** [comment]
@Mailaender: can you confirm whether this is still not working?  We should merge this if that was a false alarm.

**Comment by Mailaender** [comment]
`make nunit` works, but not the manual definitions. Probably just me doing it wrong.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e8d7538ad2f7d8ef7e3622de5c6e11f7e3aa8e6a...993d5faef6f8d70c0492f2c4997f86599629c5af)

**Pull request merged**